clippy_lints/src/unspecified_layout_union.rs

```diff
@@ -0,0 +1,105 @@
+use rustc_middle::ty::layout::LayoutOf;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+use rustc_typeck::hir_ty_to_ty;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Displays a warning when a union is declared without a `#[repr]` attribute.
+    ///
+    /// ### Why is this bad?
+    /// Unions in Rust have unspecified layout by default, despite many people thinking that they
+    /// lay out each type at the start of the union (like C does). That is, there are no guarantees
+    /// about the offset of the fields for unions with multiple non-ZST fields with not explicitly
+    /// specified layout. These cases may lead to undefined behaviour in unsafe blocks.
+    ///
+    /// ### Example
+    /// ```rust
+    /// union Foo {
+    ///     a: i32,
+    ///     b: u32,
+    /// }
+    ///
+    /// fn main() {
+    ///     let _x: u32 = unsafe {
+    ///         Foo { a: 0_i32 }.b // Undefined behaviour: `b` is allowed to be padding
+    ///     };
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// #[repr(C)]
+    /// union Foo {
+    ///     a: i32,
+    ///     b: u32,
+    /// }
+    ///
+    /// fn main() {
+    ///     let _x: u32 = unsafe {
+    ///         Foo { a: 0_i32 }.b // Now defined behaviour, this is just an i32 -> u32 transmute
+    ///     };
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub UNSPECIFIED_LAYOUT_UNION,
+    restriction,
+    "union without a `#[repr(C)]` attribute"
+}
+declare_lint_pass!(UnspecifiedLayoutUnion => [UNSPECIFIED_LAYOUT_UNION]);
+
+impl<'tcx> LateLintPass<'tcx> for UnspecifiedLayoutUnion {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if !has_c_repr_attr(cx, item.hir_id()) && is_union_with_two_non_zst_fields(cx, item) {
+            span_lint_and_help(
+                cx,
+                UNSPECIFIED_LAYOUT_UNION,
+                item.span,
+                "union which is not marked with a `#[repr(C)]` attribute",
+                None,
+                &format!(
+                    "consider annotating `{}` with `#[repr(C)]` to explicitly specify memory layout",
+                    cx.tcx.def_path_str(item.def_id.to_def_id())
+                ),
+            );
+        }
+    }
+}
+
+/// Returns true if the given item is a union with at least two non-ZST fields.
+fn is_union_with_two_non_zst_fields(cx: &LateContext<'_>, item: &Item<'_>) -> bool {
+    if let ItemKind::Union(data, _) = &item.kind {
+        data.fields().iter().filter(|f| !is_zst(cx, f.ty)).count() >= 2
+    } else {
+        false
+    }
+}
+
+fn is_zst(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>) -> bool {
+    if hir_ty.span.from_expansion() {
+        return false;
+    }
+    let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+    if let Ok(layout) = cx.layout_of(ty) {
+        layout.is_zst()
+    } else {
+        false
+    }
+}
+
+fn has_c_repr_attr(cx: &LateContext<'_>, hir_id: HirId) -> bool {
```

Comment:
I think the lint should allow any `repr`, not just `repr(C)`.

Reply:
Even if the union has some alignment attribute (like `align` or `packed`), it still has no any guarantees about the layout, if the default `Rust` representation is used. I think, those cases more likely will contain bugs, because the user may want to set a specific layout using only these attributes.

Reply:
Gotcha. Let's just add a comment somewhere that C is the only non-default representation for unions.

Reply:
We have the following description:

> ### What it does
> Displays a warning when a union is declared with the default representation (without a `#[repr(C)]` attribute)

Should we explain this in more detail?

Reply:
I think it's okay but you're welcome to expand that.

Reply:
Well, I'm not sure how best to expand this so as not to annoy the user with unnecessary repetitions. I think, it's enough, because if the user wants to restrict the language, they at least understand which layouts are possible in Rust.